draft-ietf-tls-esni.md

```diff
@@ -429,6 +429,18 @@ The "ech_outer_extensions" extension is only used for compressing the
 ClientHelloInner. It MUST NOT be sent in either ClientHelloOuter or
 ClientHelloInner.
 
+Note that it is possible to implement decoding of the EncodedClientHelloInner in
+a way that creates a denial-of-service vulnerability. Specifically, the server
+needs to check that each extension in the OuterExtensions list appears in the
+ClientHelloOuter. The naive strategy would require N\*M comparisons in the worst
+case, where N is the number of extensions in the ClientHelloOuter and M is the
+number of extensions in the OuterExtensions list. Malicious clients could
+exploit this behavior in order to degrade the performance or availability of the
+server. This problem can be mitigated by representing OuterExtensions in a way
+that allows it to be searched more quickly. for example, the worst case runtime
+can be improved to N\*log(M) by sorting the OuterExtensions and using binary
+search to access it.
```

Comment:
You want to represent the ClientHelloOuter's extensions this way, not the OuterExtensions, right? OuterExtensions is the structure stored in the outer_extensions extension. (Gosh, so many variations on "outer" and "extensions"!)

Reply:
I was thinking it would be better to pre-sort OuterExtensions, since presumably M << N.

Reply:
O(M log(M)) might instead of O(N log(N)) seems better for sorting.